### ISSUE TITLE
bug in handling of TActionResultPtrs

### ISSUE THREAD

**Issue by eguiraud** [bluehood/tdataframe#63]
There is an ownership issue with TActionResultPtrs:
we return concrete instances of them to users, instantiated on the stack, and also store them in `TDataFrameImpl` inside shared shared pointers (who think they are the sole owners of these objects).

Depending whether tail call optimization is performed or not, these two objects might or might not be the exact same.

A possible solution: we do not store `TActionResultPtr`s inside `TDataFrameImpl`, but instead only a shared pointer to their `fReady` variable (that might even be a `std::atomic_flag`).

Disadvantage of this solution: `d.Histo()->Draw()` will become impossible (now it doesn't work either, because of this bug)

**Comment by eguiraud** [comment]
errata: I had it wrong, `TDataFrameImpl` stores raw pointers to `TActionResultPtrs`, so the nature of the problem changes slightly. The following snippet can result in segfaults:

```c++
#include "TDataFrame.hxx"

int main() {
   {
      TFile wf("emptyTree.root", "RECREATE");
      TTree t("emptyTree", "emptyTree");
      int a;
      t.Branch("a", &a);
      t.Write();
   }

   TFile f("emptyTree.root");
   ROOT::TDataFrame d("emptyTree", &f);
   auto c = d.Count();
   {
      auto c2 = d.Count();
      auto c4 = d.Count();
      auto c3 = d.Count();
      auto c5 = d.Count();
      auto c6 = d.Count();
      auto c7 = d.Count();
      auto c8 = d.Count();
      auto c9 = d.Count();
      auto c10 = d.Count();
      auto c11 = d.Count();
   }
   *c; // TDataFrameImpl tries to call `SetReady` on the deleted TActionResultPtrs here
   
   return 0;
}
```

**Comment by dpiparo** [comment]
This is correct and needs to be solved: nice reproducer. I think I have a solution which can allow to get rid of the `TActionResultPtrBase` and the bug in one go: I'll update #62 with it.

**Comment by eguiraud** [comment]
fixed by pull request #62 (merged)
